### PR TITLE
INTYGFV-16624: Aggregation for questions from list controller

### PR DESCRIPTION
**Pull request by mikaelaromild1**

### CODE REVIEW

File: web/src/main/java/se/inera/intyg/webcert/web/csintegration/integration/dto/MessageQueryCriteriaDTO.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import lombok.Builder;
+import lombok.Value;
+import se.inera.intyg.common.support.facade.model.CertificateStatus;
+import se.inera.intyg.webcert.web.csintegration.integration.dto.MessageQueryCriteriaDTO.MessageQueryCriteriaDTOBuilder;
+import se.inera.intyg.webcert.web.csintegration.patient.PersonIdDTO;
+
+@JsonDeserialize(builder = MessageQueryCriteriaDTOBuilder.class)
+@Value
+@Builder
+public class MessageQueryCriteriaDTO {
```
Comment (mikaelaromild1) [inline web/src/main/java/se/inera/intyg/webcert/web/csintegration/integration/dto/MessageQueryCriteriaDTO.java:34]: We need to double check what this should contain

Comment (mhornfeldt) [reply web/src/main/java/se/inera/intyg/webcert/web/csintegration/integration/dto/MessageQueryCriteriaDTO.java:34]: It needs to be updated to match the filters of the actual list, this is for the draft list and it's not the same on all values

File: web/src/main/java/se/inera/intyg/webcert/web/service/arende/ArendeServiceImpl.java [hunk 2]
```diff
@@ -515,7 +523,6 @@ public QueryFragaSvarResponse filterArende(QueryFragaSvarParameter filterParamet
         }
 
         int originalStartFrom = filter.getStartFrom();
-        int originalPageSize = filter.getPageSize();
```
Comment (mhornfeldt) [inline web/src/main/java/se/inera/intyg/webcert/web/service/arende/ArendeServiceImpl.java:518]: Check if this is needed

File: web/src/main/java/se/inera/intyg/webcert/web/service/facade/list/ListQuestionsFacadeServiceImpl.java [hunk 3]
```diff
@@ -68,14 +75,21 @@ public ListInfo get(ListFilter filter) {
 
         final var convertedFilter = questionFilterConverter.convert(filter);
 
-        final var listResponse = arendeService.filterArende(convertedFilter, true);
-        final var convertedList = listResponse.getResults()
+        final var listResponseFromWC = arendeService.filterArende(convertedFilter, true);
+        final var listResponseFromCS = listCertificateQuestionsFromCS.list(convertedFilter.getPatientPersonId());
+
+        final var aggregatedList = Stream.concat(listResponseFromWC.getResults().stream()
+                .map(this::decorateWithResourceLinks), listResponseFromCS.getResults().stream())
```
Comment (mhornfeldt) [inline web/src/main/java/se/inera/intyg/webcert/web/service/facade/list/ListQuestionsFacadeServiceImpl.java:82]: Move the stream of WC to a variable so it's more readable

**Review by mhornfeldt — APPROVED** [review]

File: web/src/main/java/se/inera/intyg/webcert/web/csintegration/integration/ListQuestionConverter.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        final var convertedReminders = Arrays.stream(question.getReminders())
+            .map(reminder -> convertArendeListItem(certificate.get(), reminder))
+            .collect(Collectors.toList());
+
+        convertedReminders.add(convertedQuestion);
+
+        return convertedReminders;
+    }
+
+    private ArendeListItem convertArendeListItem(CertificateDTO certificate, String subject, String author, LocalDateTime sent,
```
Comment (mww-sopra-steria) [inline web/src/main/java/se/inera/intyg/webcert/web/csintegration/integration/ListQuestionConverter.java:51]: Don't we need to add status, signDate, forwarded and resourcelinks also to the ArendeListItem?

**Review by mww-sopra-steria — CHANGES_REQUESTED** [review]
Wow! Great work! It is quite complicated to aggregate questions.

I haven't been able to go through the whole PR yet, but I noticed a few things that I would like to raise so you can follow up on it straight away. I will continue the review during the day and please reach out if I misunderstand something or you would like to discuss.

File: web/src/main/java/se/inera/intyg/webcert/web/csintegration/integration/ListQuestionConverter.java (hunk 4 above)
Comment (mhornfeldt) [reply web/src/main/java/se/inera/intyg/webcert/web/csintegration/integration/ListQuestionConverter.java:51]: Which value do you mean for signDate?

Comment (mww-sopra-steria) [reply web/src/main/java/se/inera/intyg/webcert/web/csintegration/integration/ListQuestionConverter.java:51]: Sorry, ignore signDate. I think I mixed up the conversion logic in CertificateListItemConverterImpl. 😅

**Comment by mhornfeldt** [comment]
- Converted ActionLinks to ResourceLinks
- Added more fields in converter
- Filtered on QuestionStatus in WC
- Filtered on only showing unhandled questions as done in old application

File: web/src/main/java/se/inera/intyg/webcert/web/csintegration/integration/QuestionIsHandledValidator.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+ */
+
+package se.inera.intyg.webcert.web.csintegration.integration;
+
+import org.springframework.stereotype.Component;
+import se.inera.intyg.webcert.persistence.model.Status;
+import se.inera.intyg.webcert.web.web.controller.api.dto.ArendeListItem;
+
+@Component
+public class QuestionIsHandledValidator {
```
Comment (mww-sopra-steria) [inline web/src/main/java/se/inera/intyg/webcert/web/csintegration/integration/QuestionIsHandledValidator.java:27]: I don't understand how QuestionIsHandledValidator and QuestionStatusValidator works. Would be great if we could go through the implementation together. Also why do we need to have Reminders here in this class? No ArendeListItem is a reminder if we fetch them from CS.

**Review by mww-sopra-steria — CHANGES_REQUESTED** [review]
Looks good! Some minor refactoring suggested.

Would be great though to look through the filtering logic together. Not saying anything is incorrect, but I don't understand enough to assess.

File: web/src/main/java/se/inera/intyg/webcert/web/csintegration/integration/QuestionIsHandledValidator.java (hunk 5 above)
Comment (mhornfeldt) [reply web/src/main/java/se/inera/intyg/webcert/web/csintegration/integration/QuestionIsHandledValidator.java:27]: Refactored validators

**Review by mww-sopra-steria — APPROVED** [review]
Well done! Looking forward to see it in action when the CS api is ready! 🙌

**Pull request merged**